Use templates for the base types to factorize code

robUx4 opened the pull request:
A lot of the logic is the same between classes. We can factorize the code once we have factorized the storage (and initialization) types in a template.

The default value is moved to a EbmlCallbacksDefault class for the types that can have a default value. So the default value is not copied in each instance of every object.

~~This was done on top of #173 but maybe it works on its own.~~

robUx4 commented on ebml/EbmlElement.h:
```diff
@@ -220,6 +307,7 @@ class EBML_DLL_API EbmlCallbacks {
         inline EbmlElement & NewElement() const { return Create(); }
         /// is infinite/unknown size allowed
         inline bool CanHaveInfiniteSize() const { return CanInfinite; }
+        virtual const bool HasDefault() const { return false; }
```
It seems this is not possible for static instances with DLL exporting.

@mbunkus do you still use libebml/libmatroska as DLL on Windows ?

mbunkus replied:
I don't. For Windows I only link statically. On other OS I can link dynamically (which is easier to do there), or rather, on those I really don't care how it's linked. For Windows I haven't tried using DLLs for libEBML/libMatroska in a very, very long time.

robUx4 replied:
OK, I'm thinking about removing support for Windows. If you look at all the [warnings](https://github.com/Matroska-Org/libebml/actions/runs/7338935322/job/19982276669), it's clear we're not doing it right. For example we're supposed to be hiding or making public some static: https://learn.microsoft.com/en-us/cpp/cpp/using-dllimport-and-dllexport-in-cpp-classes. To do it right we should have our internal class and the version with just the bits that we want to make available to other classes. Users of our exported classes might not be using the same STL: https://stackoverflow.com/a/11189727. And it seems we can't export CRTError properly for the same reason.

We are not using it in VLC except if it's build like that on the Linux distro.

robUx4 commented:
I need to update libmatroska to build with this branch

mbunkus reviewed (changes requested):
Doesn't merge with the current master at the moment. Can you please rebase? I'll review then.

robUx4 commented:
> Doesn't merge with the current master at the moment. Can you please rebase? I'll review then.

As soon as it works in libmatroska :)

mbunkus commented on ebml/EbmlElement.h:
```diff
@@ -220,6 +277,7 @@ class EBML_DLL_API EbmlCallbacks {
         inline EbmlElement & NewElement() const { return Create(); }
         /// is infinite/unknown size allowed
         inline bool CanHaveInfiniteSize() const { return CanInfinite; }
+        virtual const bool HasDefault() const { return false; }
```
Compiler warning:

```
lib/libebml/ebml/EbmlElement.h:280:17: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
virtual const bool HasDefault() const { return false; }
```

Just change it to `virtual bool HasDefault() const` both here & in `EbmlCallbacksDefault`.

mbunkus replied:
Additional compiler warning:

```
lib/libebml/ebml/EbmlElement.h:262:20: warning: 'libebml::EbmlCallbacks' has virtual functions but non-virtual destructor [-Wnon-virtual-dtor]
class EBML_DLL_API EbmlCallbacks {                                                                                                                                                                                                                                                                                              
```

Adding a simple

```
    virtual ~EbmlCallbacks() {}
```

seems to get rid of them, but at the cost of a compiler error in various later source files, e.g. `EbmlDummy.cpp`:

```
lib/libebml/src/EbmlDummy.cpp:13:1: fatal error: constexpr variable cannot have non-literal type 'const EbmlCallbacks'
DEFINE_EBML_CLASS_ORPHAN(EbmlDummy, 0xFF, 1, "DummyElement")
^
lib/libebml/ebml/EbmlElement.h:150:60: note: expanded from macro 'DEFINE_EBML_CLASS_ORPHAN'
#define DEFINE_EBML_CLASS_ORPHAN(x,id,idl,name)            DEFINE_xxx_CLASS_ORPHAN(x,id,idl,name,GetEbmlGlobal_Context)
                                                           ^
lib/libebml/ebml/EbmlElement.h:145:38: note: expanded from macro 'DEFINE_xxx_CLASS_ORPHAN'
    constexpr const EbmlCallbacks x::ClassInfos(x::Create, Id_##x, false, name, Context_##x); \
                                     ^
lib/libebml/ebml/EbmlElement.h:274:13: note: 'EbmlCallbacks' is not literal because it has a user-provided destructor
    virtual ~EbmlCallbacks() {}
            ^
```

robUx4 replied:
I fixed both. It's odd I don't get these warnings in clang.

mbunkus replied:
Hmmm the `warning: 'libebml::EbmlCallbacks' has virtual functions but non-virtual destructor [-Wnon-virtual-dtor]` warning is still there in revision 5b434b7a011de77864de2c14a14d880a71d9afed.

robUx4 replied:
Fixed. I also added #195 to make these issues more visible in the future.

mbunkus commented:
I still get an error:

```
lib/libmatroska/src/KaxSemantic.cpp:113:1: fatal error: constexpr variable cannot have non-literal type 'const EbmlCallbacks'
DEFINE_MKX_BINARY_CONS(KaxSimpleBlock, 0xA3, 1, KaxCluster, "SimpleBlock")
^
lib/libmatroska/matroska/KaxDefines.h:30:32: note: expanded from macro 'DEFINE_MKX_BINARY_CONS'
    constexpr EbmlCallbacks a::ClassInfos(a::Create, Id_##a, false, e, Context_##a);
                               ^
lib/libebml/ebml/EbmlElement.h:274:13: note: 'EbmlCallbacks' is not literal because it has a non-trivial destructor
    virtual ~EbmlCallbacks() = default;
            ^
lib/libebml/ebml/EbmlElement.h:274:13: note: destructor for 'EbmlCallbacks' is not trivial because it is virtual
1 error generated.

```

robUx4 commented:
I forgot to push the lastest version of the libmatroska counterpart: https://github.com/Matroska-Org/libmatroska/pull/134
It should work now.

mbunkus commented:
Thanks. Compiles now. But there are a ton of issues from crashing due to null pointer access to use of uninitialized values. I've fixed some of them locally, will try to look into the others over the next couple of days. Definitely not ready for merging yet.

mbunkus reviewed (changes requested):
There are at least three issues that I found. I couldn't really fix them as I didn't understand your new structure enough. Please look into the following:

## nullptr arguments not checked

Operators & functions taking pointers are dereferencing the pointers, leading to invalid memory access or exceptions being thrown when `nullptr` is given as an argument. This immediately happens in `operator ==` in both `EbmlString` & `EbmlUnicodeString`. I don't have the stack trace at hand, but both functions were called with a `nullptr` arg, causing `std::string`/`std::wstring` to throw an `std::logic_error`.

Not sure what the right way to handle `nullptr` is. Two possibilities:

1. Always return `false`, as in "must not be used that way, cannot be equal"
2. Check if stored `Value` is empty & return `true` if so, as in "no string and empty string are the same"

Another function that comes to mind here is `EbmlElementDefaultSameStorage::IsSmallerThan`.


## reads from uninitialized values

Using valgrind I can see tons of errors of the form of making decisions based on uninitialized `Value` member variables, mostly in `EbmlUInteger`. This is unsurprising as the `EbmlElementDefaultSameStorage` and `EbmlElementDefaultStorage` don't initialize their `Value` member variable, neither statically in the class definition, nor in the constructor. The constructor in the original `EbmlUInteger` class from before the changes initialized their value member to the class's default value according to the specs, or to 0 (or the variable's equivalent) if there was no default.

This definitely affects `EbmlUInteger`, `EbmlSInteger`, `EbmlFloat`, `EbmlDate` as they have trivial member variables, which are only initialized if the code requests it.

It likely does not affect `EbmlString`, `EbmlUnicodeString`, `EbmlBinary`, `EbmlCrc32` as those use non-trivial types which do get initialized via their default constructor.


Please take a look at those issues.

robUx4 commented:
I simplified the PR to not include the code factorization to make it easier to spot the regressions. The libmatroska counterpart is still needed with this PR.

robUx4 commented:
I added some tests for strings with and without a default value and did hit some null pointer dereferencing. It was using the default value when there isn't one. It may also be the valgrind issues you saw. The default value should never be used if the default class says it has no default value.

So I think this branch should be safe now.

mbunkus commented:
I cannot test the branch in its current form. I need to be able to query an existing element if it has a default value. In 1.x I can do that via `elt.DefaultISset()`, which has been removed, which is fine. In earlier iterations of this PR I could go via the `EbmlCallbacks` with `elt.GetClassInfo().HasDefault()`, which was fine, too.

However, this current iteration doesn't have a way to retrieve the `EbmlCallbacks` anymore.

robUx4 commented:
Rebased on top of #204. I added the DefaultIsSet() call to the macro tests in #204 and turned it into the new call to use in this PR: `GetElementSpec().HasDefault()`.

robUx4 commented:
I reworked the code further. I am now testing default values on an `EbmlElement` in the tests (ie without knowing anything about the underlying class).

There is no longer a type that holds the default value if there is one. There's one `EbmlCallbacksWithDefault` template that holds the acutal default value. Types without a default value use `EbmlCallbacksDefault`.

mbunkus commented:
Thanks. With `elt.ElementSpec().HasDefault()` I can now compile this branch (together with the corresponding one for libmatroska). Unfortunately most of my test cases still fail due to different checksums (at least there are no more crashes as far as I can see; so the `nullptr` comparison seems to have been solved). For example, here's how the EBML Head looks like with a file produced by a mkvmerge compiled with v1.x:

```
+ EBML head
|+ EBML version: 1
|+ EBML read version: 1
|+ Maximum EBML ID length: 4
|+ Maximum EBML size length: 8
|+ Document type: matroska
|+ Document type version: 4
|+ Document type read version: 1
```

Here's what the same code compiled with branch `default_init/11` looks like:

```
+ EBML head
|+ EBML version: 0
|+ EBML read version: 0
|+ Maximum EBML ID length: 0
|+ Maximum EBML size length: 0
|+ Document type: matroska
|+ Document type version: 4
|+ Document type read version: 1
```

Another interesting (but wrong) effect is that the cues suddenly contain "cue codec state" elements, even if the file doesn't contain any codec state elements! They're just… there suddenly. Again I'm comparing files created with the same mkvmerge code, just compiled with different libebml/libmatroska releases.

It seems that the initial values are still not initialized. Running the mkvmerge from `default_init/11` with valgrind, I get lots of warnings such as this:

```
==986973== Conditional jump or move depends on uninitialised value(s)
==986973==    at 0x7290EC: libebml::EbmlUInteger::UpdateSize(std::function<bool (libebml::EbmlElement const&)>, bool) (lib/libebml/src/EbmlUInteger.cpp:59)
==986973==    by 0x725D44: libebml::EbmlMaster::UpdateSize(std::function<bool (libebml::EbmlElement const&)>, bool) (lib/libebml/src/EbmlMaster.cpp:112)
==986973==    by 0x7235FD: libebml::EbmlElement::Render(libebml::IOCallback&, std::function<bool (libebml::EbmlElement const&)>, bool, bool) (lib/libebml/src/EbmlElement.cpp:486)
==986973==    by 0x3FE733: render_ebml_head (src/merge/output_control.cpp:492)
==986973==    by 0x3FE733: render_headers (???:581)
==986973==    by 0x3FE733: create_next_output_file() (???:1425)
==986973==    by 0x39DDDD: main (src/merge/mkvmerge.cpp:3206)
```

I definitely do not have those warnings when running valgrind on the "good" mkvmerge compiled against v1.x.

These types of decisions based on uninitialized values could easily explain the behavior mentioned above with the "cue codec state" elements suddenly appearing.

robUx4 commented:
Thanks for all the tests. If the header is broken it's definitely something that can be tested in libebml. I'll give this a try.

Can you check that #204 works for you ? It will help reduce the scope of this PR.

mbunkus commented:
Done! #204 is fine & can be merged.

robUx4 commented:
I found the issue. The old constructor with a default value would initialize the element to its default value. IMO the element should not be initialized until it's manually done. If not it has its default value to use or should not be used at all.

For I went with the forced initialization. Only for elements with a default value, the other ones should not be used without setting a value (or their size is 0 = default value for the type).

Now it passes the test from #206.

mbunkus reviewed (approved):
Thanks! Now it passes all of my tests as well. Nice work!

Can be merged (along with the corresponding branch for libmatroska).

robUx4 commented:
Awesome. Thanks for all the testing.
